refactor(subnets): use GenericTable for ReservedRanges Table MAASENG-5280

nehjoshi opened the pull request:
## Done

- Refactored the original ```MainTable``` in ```ReservedRanges.tsx``` to ```GenericTable```.
- Created a custom hook (```useReservedRangesColumns```) for generating table columns.
- Updated ```ReservedRanges.test.tsx``` to reflect the component refactoring:
  - Some tests were incorrectly querying elements by their accessible name (e.g. using column headers like ```Labels.Type```, ```Labels.Owner``` etc.). The accessible name of a ```<td>``` should only reflect its content (e.g. for ```<td>Value here</td>``` the accessible name should be "Value here"), so the tests now assert against the cell’s actual value instead. [Read more here](https://developer.mozilla.org/en-US/docs/Glossary/Accessible_name).
  - Changed ```render(...)``` to ```renderWithProviders(...)```.

<!--
- Itemised list of what was changed by this PR.
-->

## QA steps

For subnets, first navigate to the "Address reservation" page for any subnet (e.g. ```{BASE_URL}/subnet/1/address-reservation```), and observe the "Reserved ranges" section:
- [x] Ensure a table is rendered with 6 columns (start IP, end IP, owner, type, comment & actions)
- [x] Ensure the "actions" column __cannot__ be sorted, and all other columns __can__ be sorted
- [x] Ensure that the owner is "MAAS" when the type of reserved range is "Dynamic"
- [x] Ensure that the text "No IP ranges have been reserved for this subnet." is visible, when the table is empty

For VLANs, first navigate to any VLAN (e.g. ```{BASE_URL}/vlan/5001```), and observe the "Reserved ranges" section:
- [x] Ensure a table is rendered with 7 columns (subnet, start IP, end IP, owner, type, comment & actions)
- [x] Ensure the "actions" column __cannot__ be sorted, and all other columns __can__ be sorted
- [x] Ensure that the owner is "MAAS" when the type of reserved range is "Dynamic"
- [x] Ensure that the text "No IP ranges have been reserved for this VLAN." is visible, when the table is empty

<!-- Steps for QA. -->

## Fixes

Resolves [MAASENG-5280](https://warthogs.atlassian.net/browse/MAASENG-5280)

<!-- Make sure you link the relevant Jira ticket or Launchpad bug above, if applicable. -->

## Screenshots

<!--
Attach any screenshots or videos that help illustrate or demonstrate the changes made in this PR.
-->

## Notes

<!--
(Optional)
Leave any additional notes for the reviewer here.
-->


[MAASENG-5280]: https://warthogs.atlassian.net/browse/MAASENG-5280?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

abuyukyi101198 commented on src/app/subnets/components/ReservedRanges/ReservedRanges.tsx:
```diff
@@ -175,47 +74,21 @@ const ReservedRanges = ({
-      sortKey: "subnet",
-    });
-  }
+  const columns = useReservedRangesColumns(
+    showSubnetColumn,
+    setSidePanelContent
+  );
```
Generally prefer not passing around contexts, better to call where it is used. Less props the better.
```suggestion
  const columns = useReservedRangesColumns(
    showSubnetColumn
  );
```

abuyukyi101198 commented on src/app/subnets/components/ReservedRanges/useReservedRangesColumns/useReservedRangesColumns.tsx:
```diff
@@ -0,0 +1,108 @@
+  Partial<ReservedRangesTableData>
+>;
+
+const useReservedRangesColumns = (
+  showSubnetColumn: boolean,
+  setSidePanelContent: SetSidePanelContent
+): ReservedRangesColumnsDef[] => {
+  return useMemo((): ReservedRangesColumnsDef[] => {
```
Move the context call in here.
```suggestion
const useReservedRangesColumns = (
  showSubnetColumn: boolean
): ReservedRangesColumnsDef[] => {
  const { setSidePanelContent } = useSidePanel();
  return useMemo((): ReservedRangesColumnsDef[] => {
```

abuyukyi101198 commented on src/app/subnets/components/ReservedRanges/ReservedRanges.tsx:
```diff
@@ -270,26 +143,20 @@ const ReservedRanges = ({
+        isLoading={ipRangeLoading}
+        noData={`No IP ranges have been reserved for this ${isSubnet ? "subnet" : "VLAN"}.`}
+        role="table"
+        sortBy={[{ id: "name", desc: true }]}
```
The sorting `id` must match whatever column you are sorting by. This table doesn't have a `"name"` column. Use ascending startIp to display the IPs in sequence.
```suggestion
        sortBy={[{ id: "startIp", desc: false }]}
```

nehjoshi commented:
1. Removed unnecessary context being passed as props
2. Changed table sorting to use "startIp" as the key
3. Added the ```useSidePanel()``` hook to ```useReservedRangesColumns.tsx``` instead of receiving context as a prop

abuyukyi101198 reviewed (approved):
LGTM!
